src/wallet/wallet_rpc_server.cpp

```diff
@@ -3089,6 +3089,171 @@ namespace tools
+    wal = std::move(rc.first);
+    if (!wal)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to create wallet";
+      return false;
+    }
+
+    epee::wipeable_string password = rc.second.password();
+   
+
+    bool was_deprecated_wallet = ((old_language == crypto::ElectrumWords::old_language_name) ||
+    crypto::ElectrumWords::get_is_old_style_seed(req.seed));
+
+    std::string mnemonic_language = old_language;
+    std::string m_mnemonic_language;
+    std::vector<std::string> language_list;
+    crypto::ElectrumWords::get_language_list(language_list);
+    if (mnemonic_language.empty() && std::find(language_list.begin(), language_list.end(), m_mnemonic_language) != language_list.end())
+    {
+      mnemonic_language = m_mnemonic_language;
+    }
+
+    if (((mnemonic_language.empty() || mnemonic_language == crypto::ElectrumWords::old_language_name) && was_deprecated_wallet))
+    {
+      if (was_deprecated_wallet)
+      {
+        // The user had used an older version of the wallet with old style mnemonics.
+        LOG_PRINT_L0(tools::wallet_rpc_server::tr("You had been using "
+          "a deprecated version of the wallet. Please use the new seed that we provide.\n"));
+      }
+      if (req.language.empty())
+      {
+        er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+        er.message = "Wallet is deprecated. Would provide new seed. Please specify language to use for new seed.";
+        return false;
+      }
+      mnemonic_language = req.language;
+      if (mnemonic_language.empty())
+        return {};
+    }
+
+    wal->set_seed_language(mnemonic_language);
+
+    crypto::secret_key recovery_val;
+    try
+    {
+      recovery_val = wal->generate(m_wallet_file, std::move(rc.second).password(), m_recovery_key, true, false, false);
+      LOG_PRINT_L0(tools::wallet_rpc_server::tr("Wallet has been restored.\n"));
+    }
+    catch (const std::exception& e)
+    {
+      handle_rpc_exception(std::current_exception(), er, WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR);
+      return false;
+    }
+
+    // convert rng value to electrum-style word list
+    epee::wipeable_string electrum_words;
+    crypto::ElectrumWords::bytes_to_words(recovery_val, electrum_words, mnemonic_language);
+    res.seed = electrum_words.data();
+
+    if (!wal)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to generate wallet";
+      return false;
+    }
+
+    // set blockheight if given
+    if (req.restore_height)
+    {
+      wal->set_refresh_from_block_height(req.restore_height);
+      wal->rewrite(m_wallet_file, password);
+    }
+    else
+    {
+      cryptonote::COMMAND_RPC_GET_HEIGHT::request hreq;
+      cryptonote::COMMAND_RPC_GET_HEIGHT::response hres;
+      hres.height = 0;
+      bool r = wal->invoke_http_json("/getheight", hreq, hres);
+      wal->set_refresh_from_block_height(hres.height);
+      wal->rewrite(m_wallet_file, password);
```

Comment:
That can throw, it needs a try/catch (and possibly other stuff here, safest to have the whole thing in).